### ISSUE TITLE
Only recommend packages with a compatible target

### ISSUE THREAD

**Issue by steveoh** [dotnet-outdated/dotnet-outdated#515]
If my project is targeting net6 the tool should not recommend updating to packages that only target net8 as they won't work. Instead is should provide the highest patch compatible with my target.

**Comment by mikey-t** [comment]
TL;DR Microsoft hasn't figured out how to solve this problem consistently.

It's actually an underlying limitation of dotnet and nuget. For example, if you use the built-in dotnet command:

```
dotnet list package --outdated
```

It will also list incompatible versions. It simply always shows the latest no matter what. There's a command switch that makes it seem like you can limit it to versions compatible with a particular framework:

```
// This "--framework" option is not helpful for many (almost all?) use cases
dotnet list package --outdated --framework net7.0
```

But this doesn't actually do what it seems to imply. Instead, it is used for when:
- Your project is targeting multiple frameworks
- You have conditional dependencies per framework version
- You want to run the `dotnet list package` command for only one of the target frameworks and it's conditional dependenices

Not super helpful for the vast majority of use cases...

What really throws people off is that it seems like the information exists on "package version framework compatibility", **AND** Nuget has it (see the "Frameworks" tab in the Nuget.org web UI). However, based on my reserach, it seems like it's just a guess based on:
- The target framework of the package itself
- The target framework of the dependencies of the package
- General "framework cross-compatibilities" - for example, if you have a package with a project that targets net6.0 and don't have any dependencies that are definitely not compatible with net8.0, then your package listed on Nuget.org will list it as compatible with net8.0 as a "Additional computed target framework(s)". But that doesn't actually mean it is compatible - it's just Nuget's best guess.

You can see some of the logic for how that Nuget.org "Frameworks" tab info is generated by looking at the Nuget Gallery code: https://github.com/NuGet/NuGetGallery. Some files of interest:

- https://github.com/NuGet/NuGetGallery/blob/main/src/GalleryTools/Commands/BackfillTfmMetadataCommand.cs
- https://github.com/NuGet/NuGetGallery/blob/main/src/NuGetGallery.Core/Services/AssetFrameworkHelper.cs

It's unclear why this isn't built directly into the nuget client and the dotnet cli. I would guess that the reasons are:
- It's slow to do this calculation, so they wouldn't want it to happen every time any package command happens
- At the end of the day, it's just a guess, so they don't want to be held to it

So what do we do? Here's my suggestions:
- I guess keep complaining to Microsoft to get their act together - it really wouldn't be that hard for them to include additional meta-data about framework compatibility like "definitely compatible" and "maybe compatible" framework versions in the Nuget API
- In the mean time we could try one of these:
  - Rely on the `dotnet list package --outdated --highest-minor` or this tool's `dotnet outdated -vl Major` variants and manually evaluate major version updates using the Nuget.org web UI (oof)
  - OR, try and incorporate the logic that the NuGetGallery uses to determine possible compatible frameworks into this dotnet-outdated tool
  - OR, maybe really get down in the mud and write some scripting that screen scrapes the "Frameworks" html and extracts the compatible and computed compatible frameworks... (oof)

**Comment by steveoh** [comment]
Thanks for looking into it, I tweeted to some dotnet guys to see if they have any thoughts but it's a pretty poor experience now. I appreciate your research.

**Comment by DamianEdwards** [comment]
@jondouglas @baronfel thoughts on this? I thought the NuGet service supported filtering by compatible target framework now?

**Comment by JonDouglas** [comment]
We are working to provide this framework information to the v3 api to which projects like this one or NuGet clients can consume directly.

Our end goal is to be able to compute compatibility of libraries to only suggest updates or availability of packages you can actually use and will install or update successfully. We aren't there yet.

**Comment by DamianEdwards** [comment]
Tracking issue?

**Comment by JonDouglas** [comment]
https://github.com/NuGet/NuGetGallery/issues/9627
